cc/toolchains/target_capability.bzl

```diff
@@ -0,0 +1,92 @@
+# Copyright 2024 The Bazel Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Implementation of the cc_target_capability rule."""
+
+load(
+    ":cc_toolchain_info.bzl",
+    "ArgsListInfo",
+    "FeatureConstraintInfo",
+    "FeatureInfo",
+    "FeatureSetInfo",
+    "TargetCapabilityInfo",
+)
+
+def _cc_target_capability_impl(ctx):
+    ft = FeatureInfo(
+        name = ctx.attr.feature_name or ctx.label.name,
+        label = ctx.label,
+        enabled = False,
+        args = ArgsListInfo(
+            label = ctx.label,
+            args = (),
+            files = depset(),
+            by_action = (),
+            allowlist_include_directories = depset(),
+        ),
+        implies = depset(),
+        requires_any_of = (),
+        mutually_exclusive = (),
+        external = False,
+        overridable = True,
+        overrides = None,
+        allowlist_include_directories = depset(),
+    )
+
+    # Intentionally does not provide FeatureImplyabilityInfo to prevent
+    # features from implying these kinds of rules.
+    return [
+        ft,
+        FeatureSetInfo(label = ctx.label, features = depset([ft])),
```

Comment:
As mentioned above, I'd prefer that this returns neither a FeatureInfo or a FeatureSetInfo.